### PR TITLE
JavaScript: Fix `DomMethodCallExpr.interpretsArgumentsAsHTML`.

### PR DESCRIPTION
**Pull request by xiemaisi**
The first argument to `setAttribute` is the attribute name, which is generally less sensitive than the second one, so I've changed it. Similar comments apply to the other changes.

I've run a [dist-compare](https://git.semmle.com/max/dist-compare-reports/blob/master/js/dom-fixes/report.md) (internal link) on nightly.slugs. The results are a bit mixed: while the new results on Lucifier look good, the ones on gatsby are false positives. They are, however, in test code, so wouldn't be displayed on LGTM.

I think a proper fix for these false positives would involve flow-labelifying the XSS query.

If we feel uncomfortable with the new results, I'm happy to scrap/postpone this PR.

### CODE REVIEW

**Comment by asger-semmle** [comment]
Worth landing IMO.

> I think a proper fix for these false positives would involve flow-labelifying the XSS query.

Let's talk about this offline

**Comment by esben-semmle** [comment]
How about restricting this to only flagging assignments to a library set of dangerous attribute names (such as "src", "href", "style")? That is what we do for ordinary property assignments such as `innerHtml` already.

The "data" property is interesting for an exploratory query, as that property often is read later and used unsanitized in an ordinary XSS sink.

**Comment by xiemaisi** [comment]
> How about restricting this to only flagging assignments to a library set of dangerous attribute names

Good idea in general, although it wouldn't help avoid the false positives in this case; they are all about setting `href` (to harmless values).

**Comment by xiemaisi** [comment]
Updated as suggested.

File: javascript/ql/src/semmle/javascript/security/dataflow/DOM.qll [hunk 1]
```diff
@@ -81,9 +81,15 @@ class DomMethodCallExpr extends MethodCallExpr {
...
-      or
-      name = "setAttributeNS" and argPos = 2
+      (
+        name = "setAttribute" and argPos = 1
+        or
+        name = "setAttributeNS" and argPos = 2
+      ) and
+      // restrict to potentially dangerous attributes
+      exists(string attr | attr = "href" or attr = "src" |
+        getArgument(argPos-1).getStringValue().toLowerCase() = attr
```
Comment (asger-semmle) [inline javascript/ql/src/semmle/javascript/security/dataflow/DOM.qll]: We already have `isUrlValuedAttribute` which seems to do this more generally, although it currently only works if the element type is known. Maybe we could reuse that?

Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/security/dataflow/DOM.qll]: Hm, interesting thought. That predicate does rely on knowing what the element type is, though, which in this case we don't. (I'll definitely add `action` and `formaction` here, though.)

**Pull request merged**